core/state/state_object.go

```diff
@@ -65,9 +55,20 @@ type stateObject struct {
 	trie Trie   // storage trie, which becomes non-nil on first access
 	code []byte // contract bytecode, which gets set when code is loaded
 
-	originStorage  Storage // Storage cache of original entries to dedup rewrites
-	pendingStorage Storage // Storage entries that need to be flushed to disk, at the end of an entire block
-	dirtyStorage   Storage // Storage entries that have been modified in the current transaction execution, reset for every transaction
+	originStorage  Storage // Storage entries that have been accessed within the current block
+	dirtyStorage   Storage // Storage entries that have been modified within the current transaction
+	pendingStorage Storage // Storage entries that have been modified within the current block
+
+	// needCommit tracks a set of storage entries that have been modified but
+	// not yet committed since the "last commit operation", along with their
+	// original values before mutation.
+	//
+	// Specifically, the commit will be performed after each transaction before
+	// the byzantium fork, therefore the map is already reset at the transaction
+	// boundary; however post the byzantium fork, the commit will only be performed
+	// at the end of block, this set essentially tracks all the modifications
+	// made within the block.
+	needCommit Storage
```

Comment:
Given that the previous 3 items are all `XXXStorage`, I'd recommend using `uncommittedStorage` for this one.

Reply:
Though, it's a bit weird, because it contains the "origin" values too, which also somewhat clash with *originStorage*. Does it make sense to have the origin in 2 different variables? Can't we just use this field as a map[hash]struct{} to track which slots need commit, but otherwise rely on originStorage for the actual value content?

Reply:
Unfortunately the tracked values in `origin`, `pending` and `uncommitted` have different meanings:

- origin: the value before the entire state transition of block
- pending: the latest value after mutation (e.g. after applying each transaction)
- uncommittedStorage: the value before applying the "uncommitted" mutations. 

Before the byzantium, the value being tracked in `uncommittedStorage` is the original value before the transaction; Post the byzantium, the value being tracked in `uncommittedStorage` is the original value before the block.

The reason for maintaining this value is, we want to make sure only dirty slots will be committed.